compiler/rustc_ast/src/tokenstream.rs

```diff
@@ -410,8 +410,19 @@ impl TokenStream {
         t1.next().is_none() && t2.next().is_none()
     }
 
-    pub fn map_enumerated<F: FnMut(usize, &TokenTree) -> TokenTree>(self, mut f: F) -> TokenStream {
-        TokenStream(Lrc::new(self.0.iter().enumerate().map(|(i, tree)| f(i, tree)).collect()))
+    pub fn map_enumerated_owned<F: FnMut(usize, TokenTree) -> TokenTree>(
+        mut self,
+        mut f: F,
+    ) -> TokenStream {
+        if let Some(inner) = Lrc::get_mut(&mut self.0) {
+            let owned = mem::take(inner);
+            *inner = owned.into_iter().enumerate().map(|(i, tree)| f(i, tree)).collect();
+            return self;
```

Comment:
I think putting the subsequent code in an `else` block would be better than the early return.

Alternatively, is there a way to combine the two blocks using `make_mut` somehow?

Reply:
make_mut might work. Though its allocation behavior is slightly different, doing the deep clone up-front instead of incrementally so that could have some effect.